src/include/mpir_hw_topo.h

```diff
@@ -0,0 +1,189 @@
+/* -*- Mode: C; c-basic-offset:4 ; indent-tabs-mode:nil ; -*- */
+/*
+ *  (C) 2019 by Argonne National Laboratory.
+ *      See COPYRIGHT in top-level directory.
+ *
+ */
+
+#ifndef MPIR_HW_TOPO_H_INCLUDED
+#define MPIR_HW_TOPO_H_INCLUDED
+
+/*
+ * Tracks hwloc_obj_type_t do not change ordering
+ */
+typedef enum {
+    MPIR_HWLOC_OBJ_MACHINE,
+    MPIR_HWLOC_OBJ_PACKAGE,
+    MPIR_HWLOC_OBJ_CORE,
+    MPIR_HWLOC_OBJ_PU,
+    MPIR_HWLOC_OBJ_L1CACHE,
+    MPIR_HWLOC_OBJ_L2CACHE,
+    MPIR_HWLOC_OBJ_L3CACHE,
+    MPIR_HWLOC_OBJ_L4CACHE,
+    MPIR_HWLOC_OBJ_L5CACHE,
+    MPIR_HWLOC_OBJ_L1ICACHE,
+    MPIR_HWLOC_OBJ_L2ICACHE,
+    MPIR_HWLOC_OBJ_L3ICACHE,
+    MPIR_HWLOC_OBJ_GROUP,
+    MPIR_HWLOC_OBJ_NUMANODE,
+    MPIR_HWLOC_OBJ_BRIDGE,
+    MPIR_HWLOC_OBJ_PCI_DEVICE,
+    MPIR_HWLOC_OBJ_OS_DEVICE,
+    MPIR_HWLOC_OBJ_MISC,
+    MPIR_HWLOC_OBJ_TYPE_MAX,
+    MPIR_MPICH_OBJ_DDR,         /* MPICH object for DDR memory */
+    MPIR_MPICH_OBJ_MCDRAM,      /* MPICH object for MCDRAM memory */
+    MPIR_MPICH_OBJ_DEFAULT = MPIR_MPICH_OBJ_DDR
```

Comment:
1. I don't think we should use `HWLOC` or `MPICH` in the name.  Just `MPIR_NODE_HW_OBJ_` might be sufficient.

2. If the enum type is `MPIR_node_hw_obj_type`, then the values should be `MPIR_NODE_HW_OBJ_TYPE_*`.

3. I think we are moving to the naming convention `MPIR_NODE_HW_OBJ_TYPE__MACHINE` (i.e., two underscores before the actual type).

4. I don't think `MPIR_NODE_HW_OBJ_TYPE__DEFAULT` should be a part of the enum itself.  It should be an external global variable (preferably as a part of the `MPIR_Process` structure).  Bonus points if it can be a CVAR-updatable value.

Reply:
All these sound reasonable I will address them. I like the `CVAR` option for the default.